ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Abstract/CanHardFork.hs

```diff
@@ -1,28 +1,49 @@
 {-# LANGUAGE DataKinds               #-}
 {-# LANGUAGE FlexibleContexts        #-}
 {-# LANGUAGE FlexibleInstances       #-}
+{-# LANGUAGE ScopedTypeVariables     #-}
+{-# LANGUAGE TypeApplications        #-}
+{-# LANGUAGE TypeFamilies            #-}
+{-# LANGUAGE TypeOperators           #-}
 {-# LANGUAGE UndecidableSuperClasses #-}
 
 module Ouroboros.Consensus.HardFork.Combinator.Abstract.CanHardFork (CanHardFork (..)) where
 
+import           NoThunks.Class (NoThunks)
+import           Data.Measure (BoundedMeasure)
 import           Data.SOP.Constraint
 import           Data.SOP.Functors (Product2)
 import           Data.SOP.InPairs (InPairs, RequiringBoth)
 import qualified Data.SOP.InPairs as InPairs
 import           Data.SOP.NonEmpty
+import qualified Data.SOP.Strict as SOP
 import           Data.SOP.Tails (Tails)
 import qualified Data.SOP.Tails as Tails
 import           Data.Typeable
 import           Ouroboros.Consensus.HardFork.Combinator.Abstract.SingleEraBlock
 import           Ouroboros.Consensus.HardFork.Combinator.InjectTxs
 import           Ouroboros.Consensus.HardFork.Combinator.Protocol.ChainSel
 import           Ouroboros.Consensus.HardFork.Combinator.Translation
+import           Ouroboros.Consensus.Ledger.SupportsMempool
 import           Ouroboros.Consensus.TypeFamilyWrappers
+import           Ouroboros.Network.SizeInBytes (SizeInBytes)
+
 {-------------------------------------------------------------------------------
   CanHardFork
 -------------------------------------------------------------------------------}
 
-class (All SingleEraBlock xs, Typeable xs, IsNonEmpty xs) => CanHardFork xs where
+class ( All SingleEraBlock xs
+      , Typeable xs
+      , IsNonEmpty xs
+      , BoundedMeasure (HardForkTxMeasure xs)
+      , NoThunks (HardForkTxMeasure xs)
+      ) => CanHardFork xs where
+  type HardForkTxMeasure xs
+
+  hardForkMeasureTx :: SOP.NS WrapTxMeasure xs -> HardForkTxMeasure xs
+
+  hardForkTxMeasureBytes :: proxy xs -> HardForkTxMeasure xs -> SizeInBytes
```

Comment:
This looks good :+1:; `HardForkTxMeasure` basically is the "union" of the per-era `TxMeasure`s; we chould even try to make that more explicit, but probably not worth it: 
```haskell
type        TxMeasureDims :: Type -> [Type]
type family TxMeasureDims blk

type TxMeasure blk = NP I (TxMeasureDims blk)

-- for example
type instance TxMeasureDims ByronBlock = 
  '[TxSizeInBytes]
type instance TxMeasureDims (ShelleyBlock (TPraos c) (AlonzoEra c)) =  
  '[TxSizeInBytes, ExUnits' (WithTop Natural)]
type instance TxMeasureDims (ShelleyBlock (Praos c)  (ConwayEra c)) =  
  '[TxSizeInBytes, ExUnits' (WithTop Natural), RefsScriptsSizeInBytes]

type instance TxMeasureDims (HardForkBlock xs) =
  Nub (ConcatMapTxMeasureDims xs) 
```

Reply:
Yes; a compositional approach would technically work here I think: Every tx has a measure in every era, and they're just zeros in the other eras.

However, I think that explodes operationally, pretty quickly. We'd already have, let's see... 9 constant-zeros in the size of every Conway tx, for example (Byron, Shelley, Allegra, and Mary get one, Alonzo gets 2 and Babbage gets 3)! Plus all the pointers in the heap representation, etc. I anticipate that will become a noticeable overhead over time.

So far -- we could also require each era's measure can represent the measures of previous eras (a la the ledger's `Translate` classes). If that requires overhead, then I think it'd take a _very_ long time for it to become a problem.

For now though, I did the explicit non-compositional thing in the interface (but the translate-y thing in the implementation), just to be "safe" for now.